src/freenet/node/DarknetPeerNode.java

```diff
@@ -1466,19 +1479,49 @@ public int sendFileOffer(HTTPUploadedFile file, String message) throws IOExcepti
 		return sendFileOffer(fnam, mime, message, data);
 	}
 
+        // handler for sendTextFeed
 	public void handleFproxyN2NTM(SimpleFieldSet fs, int fileNumber) {
 		String text = null;
 		long composedTime = fs.getLong("composedTime", -1);
 		long sentTime = fs.getLong("sentTime", -1);
 		long receivedTime = fs.getLong("receivedTime", -1);
-	  	try {
+		try {
 			text = Base64.decodeUTF8(fs.get("text"));
 		} catch (IllegalBase64Exception e) {
 			Logger.error(this, "Bad Base64 encoding when decoding a N2NTM SimpleFieldSet", e);
 			return;
 		}
-		N2NTMUserAlert userAlert = new N2NTMUserAlert(this, text, fileNumber, composedTime, sentTime, receivedTime);
+		long msgid = fs.getLong("msgid", Long.MAX_VALUE);
+		List<UserAlert> merged = new ArrayList<>();
+		if (msgid != Long.MAX_VALUE) {
+			// merge existing user alerts
+			// TODO: This costs time linear to the number
+			// of known alerts, so receiving a large text
+			// message is quadratic in its size!
+			for (UserAlert userAlert : node.clientCore.alerts.getAlerts()) {
```

Comment:
I think I understand what you’re doing here but I also think it doesn’t work because you only locate the alerts with `composedTime + 1` and `composedTime - 1` which may not be enough by far.

I would recommend the following:

a) break this out into a separate method.
b) get the list of alerts, filter it (both by class and msgid), then sort it by composeTime.

Now merging adjacent alerts should be a breeze (and about O(n + n lg n), if I got that right).